File: Task 1 (Shortest Queue Task)/Java Version/Queue.java
package priority_queue;

// Queue implementation 
public class Queue {
    public Human head;
    public Human tail;
    public int length;
    public int maxlength;  // Maximum Capacity of the queue
   
    // Creates empty queue of length 5
    public Queue() {
        this.head = null;
        this.tail = null;
        this.length = 0;
        this.maxlength = 5;
    }
    
    // Adds a person to the end of the queue 
    
    public void enqueue(Human human) {      // if it is empty, make him the head of the queue
        if (this.length == 0) {
            this.head = human;
            this.tail = human;
        } else {                           // Else, add it the the end 
            Human temp = this.tail;
            Human newNode = human;
            temp.next = newNode;
            this.tail = newNode;
        }
        this.length++;
    }
    
    //Removes the person at the top of the queue
    
    public void dequeue() {
        this.head = this.head.next;
        this.length--;
    }
    
   
    // check if the line is full
    public boolean full() {
        return (this.length >= this.maxlength);
    }
    
    //Empty the queue
    public void empty() {
        this.head = null;
        this.tail = null;
        this.length = 0;
    }
    
    // Inserts a person to the line based on his criticality 
    public void insert(Human type) {
        if (this.length == 0) {                                              // if the queue is empty, just enqueue 
            this.enqueue(type);
        } else {
            Human temp = this.head;
            
            if (this.head.critical_level < type.critical_level) {           //the person's criticality is higher than head, insert at front 
                type.next = this.head;
                this.head = type;
            } else {
                while (temp.next != null && type.critical_level <= temp.next.critical_level) {    // Find the right position
                    temp = temp.next;
                }
                Human temp2 = temp.next;
                temp.next = type;
                type.next = temp2;
            }
            this.length++;
        }
    }   
    
    public int numberOfpersons() {           // The current actual number of persons                    
        Human temp = this.head;
        int count = 0;
        while (temp != null) {
            count++;
            temp = temp.next;
        }
        return count;
    }
    
    public void printing() {                 // Prints the queue
        Human ptr = this.head;
        System.out.print("Queue : ");
        while (ptr != null) {
            System.out.print(ptr.type + "(" + ptr.criticality + ") ->");
            ptr = ptr.next;
        }
        System.out.println("\n");
    }
    
}

File: Task 1 (Shortest Queue Task)/Queue.py
import Humen


class Queue:
    def __init__(self):
        first_node = None
        self.head = first_node
        self.tail = first_node
        self.length = 0
        self.max_length = 5

    def enqueue(self, human):

        if self.length == 0:
            self.head = human
            self.tail = human
        else:
            ptr = self.tail
            new_node = human
            ptr.next = new_node
            self.tail = new_node
        self.length += 1

    def dequeue(self):

        self.head = self.head.next
        self.length -= 1

    def test(self):
        ptr = self.head
        count = 0
        while ptr != None:
            count += 1
            ptr = ptr.next
        return count

    def printing(self):
        ptr = self.head
        print("Queue :", end=" ")
        while ptr != None:
            print(f"{ptr.type}({ptr.criticality}) ->", end="")
            ptr = ptr.next
        print("\n")

    def full(self):
        if self.length >= self.max_length:
            return True
        return False

    def empty(self):
        self.head = None
        self.tail = None
        self.length = 0

    def insert(self, type):

        if self.length == 0:
            self.enqueue(type)
        else:
            ptr = self.head
            pregnant_woman = 4
            oldman = 3
            sick_person = 2
            normal_person = 1
            if self.head.critcial_level < type.critcial_level:
                type.next = self.head
                self.head = type
            else:
                while ptr.next != None and type.critcial_level <= ptr.next.critcial_level:
                    ptr = ptr.next
                temp = ptr.next
                ptr.next = type
                type.next = temp
            self.length += 1



